File: hestiaGO/hestiaTESTING/HasExecuted_test.go
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

package hestiaTESTING

import (
	"testing"
)

func _testHasExecutedScenarios() []*Scenario {
	return []*Scenario{
		{
			Description: `
Test HasExecuted() is able to work properly with proper Scenario settings.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test HasExecuted() is able to work properly with empty Name setting.
`,
			Switches: []string{
				cond_EMPTY_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test HasExecuted() is able to work properly with empty Switches setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_EMPTY_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test HasExecuted() is able to work properly with nil Switches setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_NIL_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test HasExecuted() is able to work properly with empty log setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_EMPTY_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test HasExecuted() is able to work properly with nil log setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_NIL_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test HasExecuted() is able to work properly with empty description setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_EMPTY_DESCRIPTION,
				cond_NIL_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test HasExecuted() is able to panic when nil Scenario is supplied.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_SUPPLY_NIL_SCENARIO,
				cond_PROPER_VERDICT,
				expect_PANIC,
			},
		}, {
			Description: `
Test HasExecuted() is able to work properly when verdict is set to unknown.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_SUPPLY_NIL_SCENARIO,
				cond_UNKNOWN_VERDICT,
				expect_PANIC,
			},
		},
	}
}

func TestHasExecutedAPI(t *testing.T) {
	scenarios := _testHasExecutedScenarios()

	for i, s := range scenarios {
		s.ID = uint64(i)
		s.Name = suite_HAS_EXECUTED_API

		// prepare
		ts := &Scenario{}
		testlib_ConfigureName(s, ts)
		testlib_ConfigureDescription(s, ts)
		testlib_ConfigureLog(s, ts)
		testlib_ConfigureSwitches(s, ts)
		testlib_ConfigureVerdict(s, ts)

		// test
		output := false
		_panick := Exec(func() any {
			if !HasCondition(s, cond_SUPPLY_NIL_SCENARIO) {
				output = HasExecuted(ts)
			} else {
				output = HasExecuted(nil)
			}
			return ""
		})
		panick := _panick.(string)

		// log output
		Log(s, _format("Test Scenario's Verdict	= %#v", ts.verdict))
		Log(s, _format("Got Output		= %v", output))
		Log(s, _format("Got Panic		= %q", panick))

		// assert
		if !testlib_AssertPanic(s, panick) {
			Conclude(s, VERDICT_FAIL)
			t.Fail()
		}

		if !assert_HasExecuted(s, output) {
			Conclude(s, VERDICT_FAIL)
			t.Fail()
		}

		if Conclusion(s) != VERDICT_FAIL {
			Conclude(s, VERDICT_PASS)
		}

		// report
		t.Logf("%s", ToString(s))
	}
}

func assert_HasExecuted(s *Scenario, output bool) bool {
	if !output {
		switch {
		case HasCondition(s, cond_SUPPLY_NIL_SCENARIO),
			HasCondition(s, cond_UNKNOWN_VERDICT):
			return true
		}
	}
	// is true

	return HasCondition(s, cond_PROPER_VERDICT)
}

File: hestiaRUST/hestia_testing/has_executed_test.rs
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.
use crate::hestia_testing;
use crate::hestia_testing::testlibs_test;

// test suites
const SUITE_NAME: &str = "hestia_testing::has_executed API";

// test conditions

// test values

// test libs
fn assert_output(s: &hestia_testing::Scenario, output: bool) -> bool {
	if hestia_testing::has_condition(s, testlibs_test::COND_PROPER_VERDICT) {
		return output;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_FAIL_VERDICT) {
		return output;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_SKIP_VERDICT) {
		return output;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_UNKNOWN_VERDICT) {
		return output;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_INITIAL_VERDICT) {
		return !output;
	}

	return false;
}

fn test_has_executed_algorithm(id: u64, desc: String, switches: Vec<String>) {
	// prepare
	let mut s: &mut hestia_testing::Scenario = &mut hestia_testing::new_scenario();
	s.id = id;
	s.name = String::from(SUITE_NAME);
	s.description = desc;
	s.switches = switches;

	// test
	let ts: &mut hestia_testing::Scenario = &mut testlibs_test::create_scenario(s);
	let output: bool = hestia_testing::has_executed(ts);

	hestia_testing::log(s, format!("Got output: '''\n{}\n'''", output));

	// assert
	hestia_testing::conclude(s, hestia_testing::VERDICT_PASS);
	if !assert_output(s, output) {
		hestia_testing::conclude(s, hestia_testing::VERDICT_FAIL);
	}

	// report
	println!("{}", hestia_testing::to_string(s));
	assert!(hestia_testing::conclusion(s) == hestia_testing::VERDICT_PASS);
}

// test suites
hestia_testing_exec!(test_has_executed_init_verdict, {
	test_has_executed_algorithm(
		4,
		"\
test hestia_testing::has_executed() is able reported executed status when set to initial value.
"
		.to_string(),
		vec![
			String::from(testlibs_test::COND_PROPER_NAME),
			String::from(testlibs_test::COND_PROPER_ID),
			String::from(testlibs_test::COND_PROPER_DESC),
			String::from(testlibs_test::COND_PROPER_SWITCHES),
			String::from(testlibs_test::COND_PROPER_LOGS),
			String::from(testlibs_test::COND_INITIAL_VERDICT),
		],
	)
});

hestia_testing_exec!(test_has_executed_unknown_verdict, {
	test_has_executed_algorithm(
		3,
		"\
test hestia_testing::has_executed() is able reported executed status when set to unknown value.
"
		.to_string(),
		vec![
			String::from(testlibs_test::COND_PROPER_NAME),
			String::from(testlibs_test::COND_PROPER_ID),
			String::from(testlibs_test::COND_PROPER_DESC),
			String::from(testlibs_test::COND_PROPER_SWITCHES),
			String::from(testlibs_test::COND_PROPER_LOGS),
			String::from(testlibs_test::COND_UNKNOWN_VERDICT),
		],
	)
});

hestia_testing_exec!(test_has_executed_skip_verdict, {
	test_has_executed_algorithm(
		2,
		"\
test hestia_testing::has_executed() is able to report executed status when set to VERDICT_SKIP.
"
		.to_string(),
		vec![
			String::from(testlibs_test::COND_PROPER_NAME),
			String::from(testlibs_test::COND_PROPER_ID),
			String::from(testlibs_test::COND_PROPER_DESC),
			String::from(testlibs_test::COND_PROPER_SWITCHES),
			String::from(testlibs_test::COND_PROPER_LOGS),
			String::from(testlibs_test::COND_SKIP_VERDICT),
		],
	)
});

hestia_testing_exec!(test_has_executed_fail_verdict, {
	test_has_executed_algorithm(
		1,
		"\
test hestia_testing::has_executed() is able to report executed status when set to VERDICT_FAIL.
"
		.to_string(),
		vec![
			String::from(testlibs_test::COND_PROPER_NAME),
			String::from(testlibs_test::COND_PROPER_ID),
			String::from(testlibs_test::COND_PROPER_DESC),
			String::from(testlibs_test::COND_PROPER_SWITCHES),
			String::from(testlibs_test::COND_PROPER_LOGS),
			String::from(testlibs_test::COND_FAIL_VERDICT),
		],
	)
});

hestia_testing_exec!(test_has_executed_proper_verdict, {
	test_has_executed_algorithm(
		0,
		"\
test hestia_testing::has_executed() is able to report executed status when set to VERDICT_PASS.
"
		.to_string(),
		vec![
			String::from(testlibs_test::COND_PROPER_NAME),
			String::from(testlibs_test::COND_PROPER_ID),
			String::from(testlibs_test::COND_PROPER_DESC),
			String::from(testlibs_test::COND_PROPER_SWITCHES),
			String::from(testlibs_test::COND_PROPER_LOGS),
			String::from(testlibs_test::COND_PROPER_VERDICT),
		],
	)
});
